File: Pinball App 2/Pinball App 2/settings/SettingsImportScreens.swift
import SwiftUI
import CoreLocation

struct AddManufacturerScreen: View {
    @ObservedObject var viewModel: SettingsViewModel
    @Environment(\.dismiss) private var dismiss
    @State private var query: String = ""
    @State private var selectedBucket: ManufacturerBucket = .modern

    private var filteredManufacturers: [PinballCatalogManufacturerOption] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let bucketed = manufacturersForSelectedBucket
        if trimmed.isEmpty {
            return bucketed
        }
        return bucketed.filter { manufacturer in
            manufacturer.name.lowercased().contains(trimmed)
        }
    }

    private var manufacturersForSelectedBucket: [PinballCatalogManufacturerOption] {
        viewModel.manufacturers.filteredForBucket(selectedBucket)
    }

    var body: some View {
        AppScreen {
            ScrollView {
                VStack(alignment: .leading, spacing: 12) {
                    VStack(alignment: .leading, spacing: 10) {
                        AppCardSubheading(text: "Bucket")

                        Menu {
                            ForEach(ManufacturerBucket.allCases) { bucket in
                                Button {
                                    selectedBucket = bucket
                                } label: {
                                    AppSelectableMenuRow(text: bucket.label, isSelected: selectedBucket == bucket)
                                }
                            }
                        } label: {
                            AppCompactDropdownLabel(text: selectedBucket.label)
                        }
                        .buttonStyle(.plain)
                    }
                    .padding(12)
                    .appPanelStyle()

                    if filteredManufacturers.isEmpty {
                        AppPanelEmptyCard(text: "No manufacturers found for that search.")
                    } else {
                        VStack(spacing: 0) {
                            ForEach(Array(filteredManufacturers.enumerated()), id: \.element.id) { index, manufacturer in
                                Button {
                                    viewModel.addManufacturer(manufacturer)
                                    dismiss()
                                } label: {
                                    SettingsImportResultRow(
                                        title: manufacturer.name,
                                        subtitle: manufacturerSubtitle(manufacturer),
                                        accessorySystemName: "plus.circle.fill",
                                        showsHighlightBadge: manufacturer.isModern,
                                        highlightBadgeText: "Modern"
                                    )
                                }
                                .buttonStyle(.plain)

                                if index < filteredManufacturers.count - 1 {
                                    AppTableRowDivider()
                                }
                            }
                        }
                        .padding(12)
                        .appPanelStyle()
                    }
                }
                .padding(.horizontal, 14)
                .padding(.vertical, 10)
            }
        }
        .searchable(text: $query, prompt: "Search manufacturers")
        .navigationTitle("Add Manufacturer")
        .navigationBarTitleDisplayMode(.inline)
    }

    private func manufacturerSubtitle(_ manufacturer: PinballCatalogManufacturerOption) -> String {
        manufacturer.gameCount == 1 ? "1 game" : "\(manufacturer.gameCount) games"
    }
}

struct AddVenueScreen: View {
    @ObservedObject var viewModel: SettingsViewModel
    @Environment(\.dismiss) private var dismiss
    @AppStorage("settings-add-venue-min-game-count") private var minimumGameCount = 5
    @State private var locationRequester = VenueLocationRequester()
    @State private var query: String = ""
    @State private var radiusMiles: Int = 25
    @State private var searchResults: [PinballLibraryVenueSearchResult] = []
    @State private var isSearching = false
    @State private var isLocating = false
    @State private var hasSearched = false
    @State private var errorMessage: String?
    @State private var lastSearchContext: String = ""

    private var filteredResults: [PinballLibraryVenueSearchResult] {
        searchResults.filter { $0.machineCount >= minimumGameCount }
    }

    private var emptyResultsMessage: String? {
        guard hasSearched else { return nil }
        if searchResults.isEmpty {
            return "No venues found for that search."
        }
        if filteredResults.isEmpty {
            return minimumGameCount == 1
                ? "No venues found with at least 1 game."
                : "No venues found with at least \(minimumGameCount) games."
        }
        return nil
    }

    private var trimmedQuery: String {
        query.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private var effectiveSearchContext: String {
        lastSearchContext.isEmpty ? trimmedQuery : lastSearchContext
    }

    private var venueSearchStatus: SettingsImportStatusContent? {
        if isLocating {
            return SettingsImportStatusContent(text: "Getting current location…", showsProgress: true)
        }
        if isSearching {
            return SettingsImportStatusContent(text: "Searching Pinball Map…", showsProgress: true)
        }
        if let errorMessage {
            return SettingsImportStatusContent(text: errorMessage, isError: true)
        }
        return nil
    }

    var body: some View {
        AppScreen {
            ScrollView {
                VStack(alignment: .leading, spacing: 12) {
                    SettingsVenueSearchControlsCard(
                        query: $query,
                        radiusMiles: $radiusMiles,
                        minimumGameCount: $minimumGameCount,
                        isSearching: isSearching,
                        isLocating: isLocating,
                        status: venueSearchStatus,
                        onSearch: {
                            Task { await runSearch() }
                        },
                        onUseCurrentLocation: {
                            Task { await runCurrentLocationSearch() }
                        }
                    )

                    if let emptyResultsMessage {
                        AppPanelEmptyCard(text: emptyResultsMessage)
                    }

                    if !filteredResults.isEmpty {
                        SettingsVenueSearchResultsPanel(
                            results: filteredResults,
                            subtitle: venueSubtitle,
                            onImport: { venue in
                                Task { await importVenue(venue) }
                            }
                        )
                    }
                }
                .padding(.horizontal, 14)
                .padding(.vertical, 10)
            }
        }
        .navigationTitle("Add Venue")
        .navigationBarTitleDisplayMode(.inline)
    }

    private func runSearch() async {
        guard !trimmedQuery.isEmpty else { return }
        await performVenueSearch(context: trimmedQuery) {
            try await PinballMapClient.searchVenues(query: trimmedQuery, radiusMiles: radiusMiles)
        }
    }

    private func runCurrentLocationSearch() async {
        isLocating = true
        do {
            let coordinate = try await locationRequester.requestCurrentLocation()
            isLocating = false
            await performVenueSearch(context: "Current location") {
                try await PinballMapClient.searchVenues(
                    latitude: coordinate.latitude,
                    longitude: coordinate.longitude,
                    radiusMiles: radiusMiles
                )
            }
        } catch {
            isLocating = false
            errorMessage = error.localizedDescription
        }
    }

    private func performVenueSearch(
        context: String,
        action: () async throws -> [PinballLibraryVenueSearchResult]
    ) async {
        hasSearched = true
        lastSearchContext = context
        isSearching = true
        defer { isSearching = false }

        do {
            searchResults = try await action()
            errorMessage = nil
        } catch {
            searchResults = []
            errorMessage = error.localizedDescription
        }
    }

    private func importVenue(_ venue: PinballLibraryVenueSearchResult) async {
        do {
            let machineIDs = try await PinballMapClient.fetchVenueMachineIDs(locationID: venueLocationID(venue))
            viewModel.importVenue(
                result: venue,
                machineIDs: machineIDs,
                searchQuery: effectiveSearchContext,
                radiusMiles: radiusMiles
            )
            dismiss()
        } catch {
            errorMessage = "Venue import failed: \(error.localizedDescription)"
        }
    }

    private func venueLocationID(_ venue: PinballLibraryVenueSearchResult) -> String {
        venue.id.replacingOccurrences(of: "venue--pm-", with: "")
    }

    private func venueSubtitle(_ venue: PinballLibraryVenueSearchResult) -> String {
        var parts: [String] = []
        if let city = venue.city, let state = venue.state {
            parts.append("\(city), \(state)")
        }
        if let distanceMiles = venue.distanceMiles {
            parts.append(String(format: "%.1f mi", distanceMiles))
        }
        parts.append(venue.machineCount == 1 ? "1 game" : "\(venue.machineCount) games")
        return parts.joined(separator: " • ")
    }
}

struct AddTournamentScreen: View {
    @ObservedObject var viewModel: SettingsViewModel
    @Environment(\.dismiss) private var dismiss
    @State private var rawTournamentID: String = ""
    @State private var isImporting = false
    @State private var errorMessage: String?

    private var tournamentID: String? {
        extractTournamentID(from: rawTournamentID)
    }

    private var canImportTournament: Bool {
        !isImporting && tournamentID != nil
    }

    var body: some View {
        AppScreen {
            ScrollView {
                VStack(alignment: .leading, spacing: 12) {
                    SettingsTournamentImportCard(
                        rawTournamentID: $rawTournamentID,
                        isImporting: isImporting,
                        errorMessage: errorMessage,
                        canImportTournament: canImportTournament,
                        onImport: {
                            Task { await importTournament() }
                        }
                    )
                }
                .padding(.horizontal, 14)
                .padding(.vertical, 10)
            }
        }
        .navigationTitle("Add Tournament")
        .navigationBarTitleDisplayMode(.inline)
    }

    private func importTournament() async {
        guard let tournamentID else { return }
        isImporting = true
        defer { isImporting = false }
        errorMessage = nil

        do {
            let tournament = try await loadTournament(id: tournamentID)
            viewModel.importTournament(result: tournament)
            dismiss()
        } catch {
            errorMessage = tournamentImportError(for: error)
        }
    }

    private func loadTournament(id: String) async throws -> MatchPlayTournamentImportResult {
        let tournament = try await MatchPlayClient.fetchTournament(id: id)
        guard !tournament.machineIDs.isEmpty else {
            throw TournamentImportError.noLinkedArenas
        }
        return tournament
    }

    private func tournamentImportError(for error: Error) -> String {
        if let tournamentImportError = error as? TournamentImportError {
            return tournamentImportError.errorDescription ?? "Tournament import failed."
        }
        return "Tournament import failed: \(error.localizedDescription)"
    }
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/settings/SettingsImportScreens.kt
package com.pillyliu.pinprofandroid.settings

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.core.content.ContextCompat
import com.pillyliu.pinprofandroid.ui.AppTintedStatusChip
import com.pillyliu.pinprofandroid.library.CatalogManufacturerOption
import com.pillyliu.pinprofandroid.library.LibraryVenueSearchResult
import com.pillyliu.pinprofandroid.ui.AnchoredDropdownFilter
import com.pillyliu.pinprofandroid.ui.AppCompactIconButton
import com.pillyliu.pinprofandroid.ui.AppCardSubheading
import com.pillyliu.pinprofandroid.ui.AppInlineTaskStatus
import com.pillyliu.pinprofandroid.ui.AppPanelEmptyCard
import com.pillyliu.pinprofandroid.ui.AppPrimaryButton
import com.pillyliu.pinprofandroid.ui.AppScreen
import com.pillyliu.pinprofandroid.ui.AppScreenHeader
import com.pillyliu.pinprofandroid.ui.CardContainer
import com.pillyliu.pinprofandroid.ui.DropdownOption
import com.pillyliu.pinprofandroid.ui.PinballThemeTokens
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

@Composable
internal fun AddManufacturerScreen(
    contentPadding: PaddingValues,
    manufacturers: List<CatalogManufacturerOption>,
    onBack: () -> Unit,
    onAdd: (CatalogManufacturerOption) -> Unit,
) {
    var query by remember { mutableStateOf("") }
    var selectedBucket by remember { mutableStateOf(ManufacturerBucket.MODERN) }
    val bucketedManufacturers = remember(manufacturers, selectedBucket) {
        manufacturers.filteredForBucket(selectedBucket)
    }
    val filtered = remember(bucketedManufacturers, query) {
        val normalized = query.trim().lowercase()
        if (normalized.isBlank()) {
            bucketedManufacturers
        } else {
            bucketedManufacturers.filter { it.name.lowercase().contains(normalized) }
        }
    }
    AppScreen(contentPadding) {
        Column(verticalArrangement = Arrangement.spacedBy(10.dp), modifier = Modifier.fillMaxSize()) {
            AppScreenHeader(title = "Add Manufacturer", onBack = onBack)
            CardContainer {
                AnchoredDropdownFilter(
                    selectedText = selectedBucket.label,
                    options = ManufacturerBucket.entries.map { bucket ->
                        DropdownOption(value = bucket.name, label = bucket.label)
                    },
                    onSelect = { value ->
                        selectedBucket = ManufacturerBucket.valueOf(value)
                    },
                    label = "Bucket",
                )
                OutlinedTextField(
                    value = query,
                    onValueChange = { query = it },
                    label = { Text("Search manufacturers") },
                    modifier = Modifier.fillMaxWidth(),
                )
            }
            if (filtered.isEmpty()) {
                AppPanelEmptyCard(text = "No manufacturers found for that search.")
            } else {
                LazyColumn(verticalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.fillMaxSize()) {
                    items(filtered) { manufacturer ->
                        CardContainer {
                            Row(verticalAlignment = Alignment.CenterVertically) {
                                Row(
                                    modifier = Modifier.weight(1f),
                                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                                    verticalAlignment = Alignment.CenterVertically,
                                ) {
                                    AppCardSubheading(manufacturer.name)
                                    if (manufacturer.isModern) {
                                        AppTintedStatusChip(
                                            text = "Modern",
                                            color = MaterialTheme.colorScheme.primary,
                                            compact = true,
                                        )
                                    }
                                    Text(
                                        "${manufacturer.gameCount} games",
                                        style = MaterialTheme.typography.bodySmall,
                                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                                    )
                                }
                                AppCompactIconButton(
                                    icon = Icons.Filled.Add,
                                    contentDescription = "Add ${manufacturer.name}",
                                    onClick = { onAdd(manufacturer) },
                                )
                            }
                        }
                    }
                }
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun AddVenueScreen(
    contentPadding: PaddingValues,
    onBack: () -> Unit,
    onImport: (LibraryVenueSearchResult, List<String>, String, Int) -> Unit,
) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val focusManager = LocalFocusManager.current
    val prefs = remember(context) { context.getSharedPreferences("settings-v1", Context.MODE_PRIVATE) }
    var query by remember { mutableStateOf("") }
    var radiusMiles by remember { mutableIntStateOf(25) }
    var minimumGameCount by remember { mutableIntStateOf(prefs.getInt("settings-add-venue-min-game-count", 5)) }
    var results by remember { mutableStateOf<List<LibraryVenueSearchResult>>(emptyList()) }
    var searching by remember { mutableStateOf(false) }
    var locating by remember { mutableStateOf(false) }
    var hasSearched by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    var lastSearchContext by remember { mutableStateOf("") }

    val filteredResults = remember(results, minimumGameCount) {
        results.filter { it.machineCount >= minimumGameCount }
    }

    fun effectiveSearchContext(): String = lastSearchContext.ifBlank { query.trim() }

    suspend fun performVenueSearch(
        contextLabel: String,
        search: suspend () -> List<LibraryVenueSearchResult>,
    ) {
        searching = true
        error = null
        hasSearched = true
        lastSearchContext = contextLabel
        runCatching { search() }
            .onSuccess {
                results = it
                error = null
            }.onFailure {
                error = it.message ?: "Venue search failed."
                results = emptyList()
            }
        searching = false
    }

    suspend fun runSearch() {
        val trimmedQuery = query.trim()
        if (trimmedQuery.isBlank()) return
        performVenueSearch(contextLabel = trimmedQuery) {
            withContext(Dispatchers.IO) { PinballMapClient.searchVenues(trimmedQuery, radiusMiles) }
        }
    }

    suspend fun runCurrentLocationSearch() {
        locating = true
        error = null
        val coordinate = runCatching { currentVenueSearchCoordinate(context) }
            .onFailure {
                error = it.message ?: "Current location is unavailable."
            }
            .getOrNull()
        locating = false
        if (coordinate == null) return

        performVenueSearch(contextLabel = "Current location") {
            withContext(Dispatchers.IO) {
                PinballMapClient.searchVenues(
                    latitude = coordinate.latitude,
                    longitude = coordinate.longitude,
                    radiusMiles = radiusMiles,
                )
            }
        }
    }

    fun launchCurrentLocationSearch() {
        if (searching || locating) return
        focusManager.clearFocus()
        scope.launch { runCurrentLocationSearch() }
    }

    val locationPermissionLauncher = rememberLauncherForActivityResult(
        contract = ActivityResultContracts.RequestMultiplePermissions(),
    ) { grants ->
        val granted = grants[Manifest.permission.ACCESS_FINE_LOCATION] == true ||
            grants[Manifest.permission.ACCESS_COARSE_LOCATION] == true
        if (granted) {
            launchCurrentLocationSearch()
        } else {
            error = "Location permission is required to search near you."
        }
    }

    fun requestCurrentLocationSearch() {
        if (searching || locating) return
        val fineGranted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_FINE_LOCATION,
        ) == PackageManager.PERMISSION_GRANTED
        val coarseGranted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_COARSE_LOCATION,
        ) == PackageManager.PERMISSION_GRANTED
        if (fineGranted || coarseGranted) {
            launchCurrentLocationSearch()
        } else {
            locationPermissionLauncher.launch(
                arrayOf(
                    Manifest.permission.ACCESS_FINE_LOCATION,
                    Manifest.permission.ACCESS_COARSE_LOCATION,
                )
            )
        }
    }

    val emptyResultsMessage = remember(hasSearched, results, filteredResults, minimumGameCount) {
        when {
            !hasSearched -> null
            results.isEmpty() -> "No venues found for that search."
            filteredResults.isEmpty() -> {
                if (minimumGameCount == 1) "No venues found with at least 1 game."
                else "No venues found with at least $minimumGameCount games."
            }
            else -> null
        }
    }

    AppScreen(contentPadding) {
        Column(verticalArrangement = Arrangement.spacedBy(10.dp), modifier = Modifier.fillMaxSize()) {
            AppScreenHeader(title = "Add Venue", onBack = onBack)
            SettingsVenueSearchCard(
                query = query,
                onQueryChange = { query = it },
                radiusMiles = radiusMiles,
                onRadiusMilesChange = { radiusMiles = it },
                minimumGameCount = minimumGameCount,
                onMinimumGameCountChange = { minimumGameCount = it },
                searching = searching,
                locating = locating,
                error = error,
                onSearch = { scope.launch { runSearch() } },
                onSearchSubmit = {
                    focusManager.clearFocus()
                    if (!searching && !locating && query.isNotBlank()) {
                        scope.launch { runSearch() }
                    }
                },
                onCurrentLocation = { requestCurrentLocationSearch() },
            )
            LaunchedEffect(minimumGameCount) {
                prefs.edit().putInt("settings-add-venue-min-game-count", minimumGameCount).apply()
            }
            emptyResultsMessage?.let {
                AppPanelEmptyCard(text = it)
            }
            LazyColumn(verticalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.fillMaxSize()) {
                items(filteredResults) { result ->
                    SettingsVenueResultCard(
                        result = result,
                        searching = searching,
                        onImport = {
                            scope.launch {
                                searching = true
                                error = null
                                runCatching {
                                    withContext(Dispatchers.IO) {
                                        PinballMapClient.fetchVenueMachineIds(result.id.removePrefix("venue--pm-"))
                                    }
                                }.onSuccess { machineIds ->
                                    onImport(
                                        result,
                                        machineIds,
                                        effectiveSearchContext(),
                                        radiusMiles,
                                    )
                                }.onFailure {
                                    error = it.message ?: "Venue import failed."
                                }
                                searching = false
                            }
                        },
                    )
                }
            }
        }
    }
}

@Composable
internal fun AddTournamentScreen(
    contentPadding: PaddingValues,
    onBack: () -> Unit,
    onImport: (MatchPlayTournamentImportResult) -> Unit,
) {
    val scope = rememberCoroutineScope()
    val focusManager = LocalFocusManager.current
    var rawTournamentId by remember { mutableStateOf("") }
    var importing by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    val tournamentId = remember(rawTournamentId) { extractTournamentId(rawTournamentId) }
    val canImportTournament = !importing && tournamentId != null

    suspend fun loadTournament(id: String): MatchPlayTournamentImportResult {
        val result = withContext(Dispatchers.IO) { MatchPlayClient.fetchTournament(id) }
        if (result.machineIds.isEmpty()) {
            throw TournamentImportException.NoLinkedArenas
        }
        return result
    }

    fun tournamentImportErrorMessage(error: Throwable): String {
        return when (error) {
            TournamentImportException.NoLinkedArenas -> error.message ?: "Tournament import failed."
            else -> error.message ?: "Tournament import failed."
        }
    }

    suspend fun performTournamentImport() {
        val resolvedId = tournamentId
        if (resolvedId == null) {
            error = "Enter a valid tournament ID."
            return
        }
        importing = true
        error = null
        runCatching { loadTournament(resolvedId) }
            .onSuccess(onImport)
            .onFailure { error = tournamentImportErrorMessage(it) }
        importing = false
    }

    AppScreen(contentPadding) {
        Column(verticalArrangement = Arrangement.spacedBy(10.dp), modifier = Modifier.fillMaxSize()) {
            AppScreenHeader(title = "Add Tournament", onBack = onBack)
            SettingsTournamentImportCard(
                rawTournamentId = rawTournamentId,
                onTournamentIdChange = { rawTournamentId = it },
                importing = importing,
                error = error,
                canImportTournament = canImportTournament,
                onImport = { scope.launch { performTournamentImport() } },
                onDone = { focusManager.clearFocus() },
            )
        }
    }
}
